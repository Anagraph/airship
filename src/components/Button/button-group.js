import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { darken } from 'polished';
import Button from './button';
import { isComponentOfType } from '../../utils';
import { colors } from '../../constants';

const separator = (props) => {
  return `1px solid ${
    !!props.secondary ? colors.primaryColor : darken(0.16, colors.primaryColor)
  }`;
};

const StyledList = styled.ul`
  border: 1px solid ${colors.primaryColor};
  border-radius: 4px;
  display: inline-flex;
  list-style: none;
  margin: 0;
  padding: 0;
  overflow: hidden;
`;

const StyledListItem = styled.li`
  border-right: ${separator};
  display: flex;

  &:last-child {
    border-right: 0;
  }
`;

class ButtonGroup extends Component {
  render() {
    const { children, secondary, ...other } = this.props;
    const buttons = React.Children.map(children, (child) => {
      return (
        <StyledListItem secondary={secondary}>
          {isComponentOfType(Button, child)
            ? React.cloneElement(child, {
                radius: 0,
                grouped: true,
                borderless: !!secondary,
                secondary: false
              })
            : child}
        </StyledListItem>
      );
    });

    return (
      <StyledList {...other} data-component="ButtonGroup">
        {buttons}
      </StyledList>
    );
  }
}

ButtonGroup.propTypes = {
  secondary: PropTypes.bool
};

export default ButtonGroup;
