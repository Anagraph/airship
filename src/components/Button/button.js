import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { darken, lighten } from 'polished';
import { colors } from '../../constants';

const font = (props) => {
  let font = "500 12px/20px 'Roboto'";
  if (props.large) {
    font = "500 16px/24px 'Roboto'";
  }
  if (props.small) {
    font = "400 10px/12px 'Roboto'";
  }
  return font;
};

const padding = (props) => {
  let padding = '6px 12px';
  if (props.large) {
    padding = '8px 16px';
  }
  if (props.small) {
    padding = '6px 8px';
  }
  return padding;
};

const background = (props) => {
  return !!props.borderless || !!props.secondary
    ? 'transparent'
    : colors.primaryColor;
};

const radius = (props) => {
  const { radius } = props;
  return radius != null && typeof radius === 'number'
    ? `${props.radius}px`
    : '4px';
};

const color = (props) => {
  return !!props.borderless || !!props.secondary
    ? colors.primaryColor
    : colors.white;
};

const border = (props) => {
  return !!props.secondary ? `1px solid ${colors.primaryColor}` : 0;
};

const hover = (props) => {
  return !!props.borderless || !!props.secondary
    ? lighten(0.45, colors.primaryColor)
    : darken(0.16, colors.primaryColor);
};

const focus = (props) => {
  return !!props.borderless || !!props.secondary
    ? lighten(0.4, colors.primaryColor)
    : darken(0.24, colors.primaryColor);
};

const StyledButton = styled.button`
  border: ${border};
  box-shadow: none;
  border-radius: ${radius};
  background: ${background};
  color: ${color};
  cursor: pointer;
  padding: ${padding};
  font: ${font};
  display: flex;
  align-items: center;

  &[disabled],
  &.is-disabled {
    opacity: 0.24;
    cursor: default;
  }

  &:hover {
    background: ${hover};
  }

  &:focus {
    background: ${focus};
    outline: none;
  }

  &.is-disabled:hover,
  &.is-disabled:focus {
    background: ${background};
    color: ${color};
  }

  .button-media {
    margin: 0 -6px;
  }

  svg + * {
    margin-left: 8px;
  }
`;

class Button extends Component {
  renderChildren() {
    const { children, grouped = false, ...others } = this.props;
    const onlyChild = React.Children.count(children) === 1;

    return React.Children.map(children, (child) => {
      // what if it's not svg?
      return typeof child === 'string' ? (
        <span>{child}</span>
      ) : (
        React.cloneElement(child, {
          className: onlyChild && !grouped ? 'button-media' : '',
          color: color(others)
        })
      );
    });
  }

  render() {
    const { children, ...other } = this.props;
    return (
      <StyledButton {...other} data-component="Button">
        {this.renderChildren()}
      </StyledButton>
    );
  }
}

Button.propTypes = {
  large: PropTypes.bool,
  small: PropTypes.bool,
  secondary: PropTypes.bool,
  borderless: PropTypes.bool,
  radius: PropTypes.number,
  grouped: PropTypes.bool
};

export default Button;
