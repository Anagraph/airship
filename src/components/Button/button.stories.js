import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { ThemeProvider } from 'styled-components';
import Button from './button';
import ButtonGroup from './button-group';
import Icon from '../Icon/icon';
import { theme } from '../../constants';

const CUSTOM_THEME = {
  ...theme,
  brand01: '#FABADA',
};

storiesOf('Button', module)
  .add('Default', () => (
    <div>
      <h3 className="header">Primary</h3>
      <Button>Click me</Button>

      <h3 className="header">Scondary</h3>
      <Button secondary>Click me</Button>

      <h3 className="header">Borderless</h3>
      <Button borderless>Click me</Button>

      <h3 className="header">With event</h3>
      <Button onClick={action('click')}>Click me</Button>

      <h3 className="header">Grouped</h3>
      <ButtonGroup>
        <Button>
          <Icon icon="plus" />
        </Button>
        <Button>Edit</Button>
        <Button>
          Delete
        </Button>
      </ButtonGroup>

      <h3 className="header">Grouped Secondary</h3>
      <ButtonGroup secondary>
        <Button>
          <Icon icon="plus" />
        </Button>
        <Button>Edit</Button>
        <Button>
          Delete
        </Button>
      </ButtonGroup>
    </div>
  ))
  .add('Disabled', () => (
    <div>
      <h3 className="header">Primary</h3>
      <Button disabled>Click me</Button>

      <h3 className="header">Scondary</h3>
      <Button disabled secondary>Click me</Button>

      <h3 className="header">Borderless</h3>
      <Button disabled borderless>Click me</Button>

      <h3 className="header">With event</h3>
      <Button disabled onClick={action('click')}>Click me</Button>

      <h3 className="header">Grouped</h3>
      <ButtonGroup>
        <Button disabled>
          <Icon icon="plus" />
        </Button>
        <Button disabled>Edit</Button>
        <Button disabled>
          Delete
        </Button>
      </ButtonGroup>

      <h3 className="header">Grouped Secondary</h3>
      <ButtonGroup secondary>
        <Button disabled>
          <Icon icon="plus" />
        </Button>
        <Button disabled>Edit</Button>
        <Button disabled>
          Delete
        </Button>
      </ButtonGroup>
    </div>
  ))
  .add('With custom theme', () => (
    <ThemeProvider theme={CUSTOM_THEME}>
      <div>
        <h3 className="header">Primary</h3>
        <Button>Click me</Button>

        <h3 className="header">Scondary</h3>
        <Button secondary>Click me</Button>

        <h3 className="header">Borderless</h3>
        <Button borderless>Click me</Button>

        <h3 className="header">With event</h3>
        <Button onClick={action('click')}>Click me</Button>

        <h3 className="header">Grouped</h3>
        <ButtonGroup>
          <Button>
            <Icon icon="plus" />
          </Button>
          <Button>Edit</Button>
          <Button>
            Delete
          </Button>
        </ButtonGroup>

        <h3 className="header">Grouped Secondary</h3>
        <ButtonGroup secondary>
          <Button>
            <Icon icon="plus" />
          </Button>
          <Button>Edit</Button>
          <Button>
            Delete
          </Button>
        </ButtonGroup>
      </div>
    </ThemeProvider>
  ));
