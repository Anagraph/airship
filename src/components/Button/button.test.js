import React from 'react';
import renderer from 'react-test-renderer';
import Button from './button';

describe('render', () => {
  it('renders without crashing', () => {
    const component = renderer.create(<Button>Button</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with large param', () => {
    const component = renderer.create(<Button large>Button</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with small param', () => {
    const component = renderer.create(<Button small>Button</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders secondary button', () => {
    const component = renderer.create(<Button secondary>Button</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders borderless button', () => {
    const component = renderer.create(<Button borderless>Button</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
