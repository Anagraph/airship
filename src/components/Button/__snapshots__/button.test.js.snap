// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders borderless button 1`] = `
<button
  className="sc-bdVaJa hSmgSf"
  data-component="Button"
>
  <span>
    Button
  </span>
</button>
`;

exports[`render renders secondary button 1`] = `
<button
  className="sc-bdVaJa gneWQO"
  data-component="Button"
>
  <span>
    Button
  </span>
</button>
`;

exports[`render renders with large param 1`] = `
<button
  className="sc-bdVaJa PwUmn"
  data-component="Button"
>
  <span>
    Button
  </span>
</button>
`;

exports[`render renders with small param 1`] = `
<button
  className="sc-bdVaJa kssjaf"
  data-component="Button"
>
  <span>
    Button
  </span>
</button>
`;

exports[`render renders without crashing 1`] = `
<button
  className="sc-bdVaJa filjFa"
  data-component="Button"
>
  <span>
    Button
  </span>
</button>
`;
